okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/zip.kt

```diff
@@ -231,298 +175,211 @@ private fun buildIndex(entries: List<ZipEntry>): Map<Path, ZipEntry> {
-  return EocdRecord(numEntries, centralDirOffset, commentLength)
-}
-
-/**
- * On exit, [this@readEntry] will be positioned at the start of the next entry in the Central Directory.
- */
+/** When this returns, [this] will be positioned at the start of the next entry. */
 @ExperimentalFileSystem
 @Throws(IOException::class)
-internal fun BufferedSource.readEntry(isZip64: Boolean): ZipEntry {
-  val sig = readIntLe()
-  if (sig.toLong() != CENSIG) {
-    throwZipException("Central Directory Entry", sig)
+internal fun BufferedSource.readEntry(): ZipEntry {
+  val signature = readIntLe()
+  if (signature != CENTRAL_FILE_HEADER_SIGNATURE) {
+    throw IOException(
+      "bad zip: expected ${CENTRAL_FILE_HEADER_SIGNATURE.hex} but was ${signature.hex}"
+    )
   }
-  skip(4)
-  val gpbf = readShortLe().toInt() and 0xffff
-  if (gpbf and GPBF_UNSUPPORTED_MASK != 0) {
-    throw IOException("Invalid General Purpose Bit Flag: $gpbf")
+
+  skip(4) // version made by (2) + version to extract (2).
+  val bitFlag = readShortLe().toInt() and 0xffff
+  if (bitFlag and BIT_FLAG_UNSUPPORTED_MASK != 0) {
+    throw IOException("unsupported zip: general purpose bit flag=${bitFlag.hex}")
   }
 
   val compressionMethod = readShortLe().toInt() and 0xffff
   val time = readShortLe().toInt() and 0xffff
-  val modDate = readShortLe().toInt() and 0xffff
+  val date = readShortLe().toInt() and 0xffff
+  // TODO(jwilson): decode NTFS and UNIX extra metadata to return better timestamps.
+  val lastModifiedAtMillis = dosDateTimeToEpochMillis(date, time)
 
   // These are 32-bit values in the file, but 64-bit fields in this object.
   val crc = readIntLe().toLong() and 0xffffffffL
-  val compressedSize = readIntLe().toLong() and 0xffffffffL
-  val size = readIntLe().toLong() and 0xffffffffL
-  val nameLength = readShortLe().toInt() and 0xffff
-  val extraLength = readShortLe().toInt() and 0xffff
+  var compressedSize = readIntLe().toLong() and 0xffffffffL
+  var size = readIntLe().toLong() and 0xffffffffL
+  val nameSize = readShortLe().toInt() and 0xffff
+  val extraSize = readShortLe().toInt() and 0xffff
   val commentByteCount = readShortLe().toInt() and 0xffff
 
-  // This is a 32-bit value in the file, but a 64-bit field in this object.
-  skip(8)
-  val localHeaderRelOffset = readIntLe().toLong() and 0xffffffffL
-  val name = readUtf8(nameLength.toLong())
-  for (element in name) {
-    if (element == '\u0000') throw IOException("Filename contains NUL byte")
+  skip(8) // disk number start (2) + internal file attributes (2) + external file attributes (4).
+  var offset = readIntLe().toLong() and 0xffffffffL
+  val name = readUtf8(nameSize.toLong())
+  if ('\u0000' in name) throw IOException("bad zip: filename contains 0x00")
+
+  val requiredZip64ExtraSize = (if (size == MAX_ZIP_ENTRY_AND_ARCHIVE_SIZE) 8L else 0L) +
+    (if (compressedSize == MAX_ZIP_ENTRY_AND_ARCHIVE_SIZE) 8L else 0L) +
+    (if (offset == MAX_ZIP_ENTRY_AND_ARCHIVE_SIZE) 8L else 0L)
```

Comment:
Could we make that more readable? Something like that maybe
```suggestion
  val requiredZip64ExtraSize = run {
    var result = 0L
    if (size == MAX_ZIP_ENTRY_AND_ARCHIVE_SIZE) result += 8
    if (compressedSize == MAX_ZIP_ENTRY_AND_ARCHIVE_SIZE) result += 8
    if (offset == MAX_ZIP_ENTRY_AND_ARCHIVE_SIZE) result += 8
    return@run result
  }
```